modules/imgcodecs/src/grfmt_webp.hpp

```diff
@@ -61,17 +27,23 @@ class WebPDecoder CV_FINAL : public BaseImageDecoder
 
     bool readData( Mat& img ) CV_OVERRIDE;
     bool readHeader() CV_OVERRIDE;
+    bool nextPage() CV_OVERRIDE;
 
     size_t signatureLength() const CV_OVERRIDE;
     bool checkSignature( const String& signature) const CV_OVERRIDE;
 
     ImageDecoder newDecoder() const CV_OVERRIDE;
 
 protected:
+
+    struct UniquePtrDeleter {
+        void operator()(WebPAnimDecoder* decoder) const { WebPAnimDecoderDelete(decoder); }
```

Comment:
Please add the same thing for data and encoder:
```
void operator()(WebPData* data) const { WebPDataClear(data); }
```
And you can then use it internally when creating a unique_ptr (or use a typedef to that)
```
typedef std::unique_ptr<WebPAnimDecoder, UniquePtrDeleter> AnimDecoder;
```
If you only use those once, no need indeed to create the UniquePtrDeleter and just do what you already did for the encoder.

Reply:
Defining `operator()` forces you to include the WebP headers here. Can you instead define it in the `.cpp` file ? And therefore forward declare `WebPAnimDecoder` in the header